### PR TITLE
Add semantics for struct type and value literals.

### PR DESCRIPTION
**Pull request by jonmeow**
This handles the basics of type and value for structs. Structurally, these look like parameters and arguments (respectively) because expressions/generics may result in multiple IR nodes being generated.

Because `{}` needs to be cast to a type for storage, I'm also adding some validation that's not specific to `{}`, e.g. that `1` shouldn't be valid as a type for storage (previously, nothing errored for that).

This adds more stringification of types, particularly literals, because they come up in value errors now.

ImplicitAs is the result of me mulling whether I'm taking the right approach on type conversions. I think it needs to return a value so that if the implicit cast rewrites the value, the result is accessible to the caller. I may reorient the current TryTypeConversion logic to be more based on the ImplicitAs logic.

### CODE REVIEW

**Comment by jonmeow** [comment]
Since there isn't an active review, I'm going to draft to work on the empty struct issue.

File: toolchain/semantics/semantics_ir.cpp [hunk 1]
```diff
@@ -128,6 +128,22 @@ auto SemanticsIR::StringifyNodeImpl(llvm::raw_ostream& out,
 
   auto node = GetNode(node_id);
   switch (node.kind()) {
+    case SemanticsNodeKind::StructType: {
+      out << "{";
+      auto refs = GetNodeBlock(node.GetAsStructType().second);
+      llvm::ListSeparator sep;
+      for (const auto& ref_id : refs) {
+        out << sep;
+        StringifyNodeImpl(out, ref_id);
```
Comment (chandlerc) [inline toolchain/semantics/semantics_ir.cpp:138]: A long way from mattering, but maybe worth leaving a todo for when we end up wanting a fuzzer to not waste time here...

```suggestion
        // TODO: Bound recursion depth or remove recursive step.
        StringifyNodeImpl(out, ref_id);
```

File: toolchain/semantics/semantics_parse_tree_handler.cpp [hunk 2]
```diff
@@ -232,68 +232,123 @@ auto SemanticsParseTreeHandler::TryTypeConversionOnArgs(
...
-    auto parse_kind = parse_tree_->node_kind(node_stack_.PeekParseNode());
-    if (parse_kind == start_kind) {
-      return on_start(node_block_stack_.Pop(), params_or_args_stack_.Pop());
-    } else if (parse_kind == comma_kind) {
-      node_stack_.PopAndDiscardSoloParseNode(comma_kind);
-    } else {
-      node_stack_.PopAndIgnore();
-    }
+  if (value_type.kind() != as_type.kind() ||
+      as_type.kind() != SemanticsNodeKind::StructType) {
```
Comment (chandlerc) [inline toolchain/semantics/semantics_parse_tree_handler.cpp]: Minor thing, but I find it a bit easier to read as:
```suggestion
  if (value_type.kind() != SemanticsNodeKind::StructType ||
      as_type.kind() != SemanticsNodeKind::StructType) {
```

File: toolchain/semantics/semantics_parse_tree_handler.cpp [hunk 3]
```diff
@@ -232,68 +232,123 @@ auto SemanticsParseTreeHandler::TryTypeConversionOnArgs(
...
-  // invalidate the ir reference.
-  auto& params = semantics_->GetNodeBlock(params_or_args_stack_.PeekForAdd());
+  for (int i = 0; i < static_cast<int>(value_type_refs.size()); ++i) {
+    auto value_type_field = semantics_->GetNode(value_type_refs[i]);
+    auto as_type_field = semantics_->GetNode(as_type_refs[i]);
+    if (value_type_field.type_id() != as_type_field.type_id() ||
+        value_type_field.GetAsStructTypeField() !=
+            as_type_field.GetAsStructTypeField()) {
+      return false;
+    }
```
Comment (chandlerc) [inline toolchain/semantics/semantics_parse_tree_handler.cpp:305]: I assume you're thinking about it and largely plan to come back to this in a subsequent step, but I'd imagine this eventually just does a transitive implicit-as logic across the tree...

Anyways, totally happy with the current logic for this iteration, and either the above or something else when you get to it.

**Review by chandlerc — APPROVED** [review]
Nice! Some minor tweaks inline and a thought that I just wanted to say "out loud" -- that one's not blocking anything.

File: toolchain/semantics/semantics_parse_tree_handler.cpp (hunk 3 above)
Comment (jonmeow) [reply toolchain/semantics/semantics_parse_tree_handler.cpp:305]: Correct, I'm working iteratively.

**Pull request merged**